core/trino-main/src/main/java/io/trino/sql/planner/SymbolAllocator.java

```diff
@@ -71,8 +71,7 @@ public Symbol newSymbol(String nameHint, Type type, @Nullable String suffix)
         requireNonNull(nameHint, "nameHint is null");
         requireNonNull(type, "type is null");
 
-        // TODO: workaround for the fact that QualifiedName lowercases parts
-        nameHint = nameHint.toLowerCase(ENGLISH);
+        nameHint = nameHint.toLowerCase(ENGLISH).replaceAll("[^a-z0-9$]+", "_");
```

Comment:
- we should also remove repeated `_` characters (if present)
- not sure about regexp overhead. CharMatcher is more verbose, but might be faster, and we are here frequently

Reply:
I'll look into that.